File: src/25/python/solutions.py
from lib.lists.linkedlists import SinglyLinkedList
from nodes import ListNode


class Solution1:
    # Iterative
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        # Initialize
        dummy: ListNode = ListNode();
        dummy.next = head
        # Type inferred
        pre, cur = dummy, head

        while True:
            for _ in range(k):
                if not cur: return dummy.next
                cur = cur.next

            pre.next = SinglyLinkedList(head).reverse_before(cur)
            pre, head = head, cur


class Solution2:
    # Recursion
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        cur: ListNode = head
        for _ in range(k):
            if not cur: return head
            cur = cur.next

        new_head: ListNode = SinglyLinkedList(head).reverse_before(cur)

        head.next = self.reverseKGroup(cur, k)
        return new_head

File: 25/cpp/solutions.cpp
#include "lib/lists/linkedlists.h"
#include "nodes.h"

using namespace lib;
using namespace nodes;

class Solution1
{
    // Ierative
public:
    ListNode *reverseKGroup(ListNode *head, int k)
    {
        ListNode *dummy = new ListNode();
        dummy->next = head;
        ListNode *pre = dummy, *cur = head;
        while (true)
        {
            for (int i = 0; i < k; i++)
            {
                if (!cur)
                    return dummy->next;
                cur = cur->next;
            }

            SinglyLinkedList ls(head);
            pre->next = ls.reverseBefore(cur);
            pre = head;
            head = cur;
        }
    }
};

class Solution2
{
    // Recursive
public:
    ListNode *reverseKGroup(ListNode *head, int k)
    {
        ListNode *cur = head;
        for (int i = 0; i < k; ++i)
        {
            if (!cur)
                return head;
            cur = cur->next;
        }

        SinglyLinkedList ls(head);
        ListNode *newHead = ls.reverseBefore(cur);
        head->next = reverseKGroup(cur, k);
        return newHead;
    }
};